File: src/nios/knowledge/policyEngine.ts
/**
 * Policy Engine — Phase 1 stub
 * Executable policies evaluated in verify() stage.
 */

export interface PolicyViolation {
  policyId: string;
  message: string;
  severity: "block" | "warn";
}

export interface PolicyContext {
  capabilityId: string;
  hasEngineEvidence: boolean;
  hasLegalCitation: boolean;
  amountMentioned: boolean;
  numericFromLlm: boolean;
}

const AI_POLICIES = [
  {
    id: "policy.ai.no_money_without_engine",
    check: (ctx: PolicyContext) =>
      ctx.amountMentioned && !ctx.hasEngineEvidence && ctx.capabilityId.includes("tax"),
    message: "Tax amounts must come from deterministic engine, not LLM",
    severity: "block" as const,
  },
  {
    id: "policy.ai.legal_requires_citation",
    check: (ctx: PolicyContext) =>
      ctx.capabilityId.includes("legal") && !ctx.hasLegalCitation,
    message: "Legal answers require graph citation",
    severity: "block" as const,
  },
];

export function evaluatePolicies(ctx: PolicyContext): PolicyViolation[] {
  const violations: PolicyViolation[] = [];
  for (const policy of AI_POLICIES) {
    if (policy.check(ctx)) {
      violations.push({
        policyId: policy.id,
        message: policy.message,
        severity: policy.severity,
      });
    }
  }
  return violations;
}

export function policiesPass(ctx: PolicyContext): boolean {
  return evaluatePolicies(ctx).every((v) => v.severity !== "block");
}

File: erp_bot/src/nios/knowledge/policy_engine.py
"""Policy Engine v1 — PolicyDSL in verify() stage."""

from __future__ import annotations

from dataclasses import dataclass

from .policy_dsl import BOOTSTRAP_POLICIES, evaluate_policy_rules


@dataclass
class PolicyContext:
    capability_id: str
    has_engine_evidence: bool = False
    has_legal_citation: bool = False
    amount_mentioned: bool = False
    amount: float = 0.0
    confidence: float = 1.0
    jurisdiction: str = "NP"
    log_output: bool = False


def evaluate_policies(ctx: PolicyContext) -> list[dict[str, str]]:
    dsl_ctx = {
        "capability_id": ctx.capability_id,
        "has_engine_evidence": ctx.has_engine_evidence,
        "has_legal_citation": ctx.has_legal_citation,
        "amount_mentioned": ctx.amount_mentioned,
        "amount": ctx.amount,
        "confidence": ctx.confidence,
        "jurisdiction": ctx.jurisdiction,
        "log_output": ctx.log_output,
    }
    return evaluate_policy_rules(BOOTSTRAP_POLICIES, dsl_ctx)


def policies_pass(ctx: PolicyContext) -> bool:
    return not any(v["severity"] == "block" for v in evaluate_policies(ctx))
